### PR TITLE
[IOTDB-4335] Implement PathPatternTree to support trigger and delete

### PR DESCRIPTION
**Pull request by Cpaulyz**
## Description
* Implement `PatternTreeMap` to support trigger and delete
    * Test and example in `PatternTreeMapTest`
    * Trigger modules can create `PathPatternTree`  directly from `PatternTreeMapFactory`
* Move `PathPatternTree` to node-commons module
* Requirement docs: https://apache-iotdb.feishu.cn/docx/doxcnF9w5N6vxnQ9qak5zyMwoig

### CODE REVIEW

File: node-commons/src/main/java/org/apache/iotdb/commons/path/PatternTreeMap.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+   * @param resultList result list
+   * @param fromMultiWildCard true if node in caller is '**', so there is no need to traverse the
+   *     remaining pathNodes
+   */
+  public void searchOverlapped(
+      PathPatternNode<V> node,
+      String[] pathNodes,
+      int pos,
+      List<V> resultList,
+      boolean fromMultiWildCard) {
```
Comment (JackieTien97) [inline node-commons/src/main/java/org/apache/iotdb/commons/path/PatternTreeMap.java:142]: `fromMultiWildCard` means there may exist two consecutive `**` in `PathPatternTree`?
I think this case should be deal with or avoided at the very beginning like during constructing this `PathPatternTree`.

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: node-commons/src/main/java/org/apache/iotdb/commons/path/PatternTreeMap.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Supplier;
+
+@NotThreadSafe
+public class PatternTreeMap<V> {
+  private final PathPatternNode<V> root;
+  private final Supplier<? extends Set<V>> supplier;
+  private BiConsumer<V, Set<V>> appendFunction;
+  private BiConsumer<V, Set<V>> deleteFunction;
```
Comment (JackieTien97) [inline node-commons/src/main/java/org/apache/iotdb/commons/path/PatternTreeMap.java:36]:
```suggestion
  private final BiConsumer<V, Set<V>> appendFunction;
  private final BiConsumer<V, Set<V>> deleteFunction;
```

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: node-commons/src/main/java/org/apache/iotdb/commons/path/PatternTreeMap.java (hunk 1 above)
Comment (Cpaulyz) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/PatternTreeMap.java:142]: `fromMultiWildCard` is unnecessary. I will remove it later.

File: node-commons/src/main/java/org/apache/iotdb/commons/path/PathPatternTree.java [hunk 3]
```diff
@@ -142,7 +141,7 @@ public List<String> getAllDevicePatterns() {
   }
 
   private void searchDevicePattern(
-      PathPatternNode curNode, List<String> nodes, Set<String> results) {
+      PathPatternNode<Void> curNode, List<String> nodes, Set<String> results) {
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/PathPatternTree.java:144]: I didn't change the semantics of this interface. It originally seemed to assume that the device of `root.sg1.d1.**` is `root.sg1.d1.**`.  I didn't know if the modification will have an impact on the upper layer application

File: node-commons/src/main/java/org/apache/iotdb/commons/path/PathPatternNode.java [hunk 4]
```diff
@@ -26,46 +26,96 @@
...
-public class PathPatternNode {
+import static org.apache.iotdb.commons.conf.IoTDBConstant.MULTI_LEVEL_PATH_WILDCARD;
+import static org.apache.iotdb.commons.conf.IoTDBConstant.ONE_LEVEL_PATH_WILDCARD;
+
+public class PathPatternNode<V> {
 
   private final String name;
-  private final Map<String, PathPatternNode> children;
+  private final Map<String, PathPatternNode<V>> children;
+  private Set<V> valueSet;
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/PathPatternNode.java:46]: > It seems that you forgot about serializing this field.

Nope. I reuse the data structure of the nodes in PathPatternTree here, mounting an additional valueSet. Serialization methods are only used in PathPatternTree, and the serialization information is used as a ByteBuffer in RPC. valueSet is always empty in PathPatternTree, so I did not serialize it to save network cost .


> And equalsWIth may also take it into account

I've added the valueSet comparison in `equalsWith`

File: node-commons/src/main/java/org/apache/iotdb/commons/path/PatternTreeMap.java (hunk 2 above)
Comment (Cpaulyz) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/PatternTreeMap.java]: done

File: node-commons/src/main/java/org/apache/iotdb/commons/path/PathPatternTree.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (JackieTien97) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/PathPatternTree.java:144]: You can discuss with @liuminghui233 and @MarcosZyk about it, I think it's a bug in previous implementation, and can be fixed by the way in this pr.

**Review by JackieTien97 — APPROVED** [review]
Excellent job!

File: node-commons/src/main/java/org/apache/iotdb/commons/path/PathPatternTree.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/PathPatternTree.java]: The device of ```root.sg1.d1.**``` may be  ```root.sg1.d1``` and ```root.sg1.d1.**```.

**Pull request merged**